Handle unique validation conflicts caused by soft-deleted records

mohammadrasoulasghari opened the pull request:
Hi Laravel Team 👋,

During my experience working on various projects, I’ve encountered a recurring issue when validating the uniqueness of a field while using Soft Deletes. Specifically, when using Rule::unique for validation, it often leads to unintended conflicts because soft-deleted records are still considered in the validation check. This has caused confusion and bugs in projects where certain records are logically “deleted” but still exist in the database.

To address this issue, I’ve introduced a new method called withSoftDeletes to the Rule::unique class. This method allows developers to exclude soft-deleted records from the unique validation check. By explicitly opting in to this behavior, it ensures that the framework remains opinionated and avoids introducing any breaking changes.

I believe this addition can improve the developer experience by:

1. “**Explicit is better than implicit**” — This method makes the behavior clear and opt-in, avoiding unexpected side effects.
2. “**Don’t repeat yourself (DRY)**” — It reduces the repetitive task of manually adding conditions for soft-deleted records.

**Usage Example:**
Here’s an example of how this method can be used:
```
use Illuminate\Validation\Rule;

$request->validate([
    'email' => [
        'required',
        'email',
        Rule::unique('users', 'email')->withSoftDeletes(),
    ],
]);
```

Thank you for considering this PR! 😊

Looking forward to your feedback and suggestions.

mohammadrasoulasghari committed:
feat(Validation): add withSoftDeletes method to Unique rule

Allows excluding soft-deleted records during unique validation.

antonkomarev commented on src/Illuminate/Validation/Rules/Unique.php:
```diff
@@ -73,4 +73,18 @@ public function __toString()
+    public function withSoftDeletes()
+    {
+        $this->where(function ($query) {
+            $query->whereNull('deleted_at');
```
This field may be named differently in each model.

mohammadrasoulasghari committed:
feat: support customizable soft delete column in unique validation

Ensure the soft delete column can be dynamically determined based on the model.

shaedrich commented:
Doesn't `DatabaseRule` have a `withoutTrashed()` method?

https://github.com/laravel/framework/blob/122e5c2a9ac6680bd0e2a4fe4db6756f01393dcb/src/Illuminate/Validation/Rules/DatabaseRule.php#L179-L190

shaedrich commented on src/Illuminate/Validation/Rules/Unique.php:
```diff
@@ -75,14 +75,19 @@ public function __toString()
+            $query->whereNull($deletedAtColumn);
         });
 
         return $this;
```
You could use `tap()` here:
```suggestion
        return tap($this)->where(function ($query) use ($model) {
            $deletedAtColumn = $model && method_exists($model, 'getDeletedAtColumn')
                ? $model->getDeletedAtColumn()
                : 'deleted_at';

            $query->whereNull($deletedAtColumn);
        });
```

shaedrich commented on src/Illuminate/Validation/Rules/Unique.php:
```diff
@@ -75,14 +75,19 @@ public function __toString()
      * @return $this
      */
-    public function withSoftDeletes()
+    public function withSoftDeletes(?Model $model = null): static
```
You might want to add similar functionality to `ignore()`, where what is passed can be one of a few options:
https://github.com/laravel/framework/blob/122e5c2a9ac6680bd0e2a4fe4db6756f01393dcb/src/Illuminate/Validation/Rules/Unique.php#L34-L44

In our case, this would be either a model or the column name directly.

mohammadrasoulasghari replied:
Thanks for your suggestions
Considering the existence of the withoutTrashed method and the fact that I was not aware of its existence, there is no need for this method and i close this pr
